doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,329 @@
+lookups at publish time, enabling very fast publish speeds and allowing
+for heavy use while incurring very minimal cost. Channels are created with
+[`diagnostics_channel.channel(name)`][], constructing a channel directly
+with `new Channel(name)` is not supported.
+
+#### `channel.hasSubscribers`
+
+* Returns: {boolean} If there are active subscribers
+
+Check if there are active subscribers to this channel. This is helpful if
+the message you want to send might be expensive to prepare.
+
+This API is optional but helpful when trying to publish messages from very
+performance-senstive code.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+if (channel.hasSubscribers) {
+  // There are subscribers, prepare and publish message
+}
+```
+
+#### `channel.publish(message)`
+
+* `message` {any} The message to send to the channel subscribers
+
+Publish a message to any subscribers to the channel. This will trigger
+message handlers synchronously so they will execute within the same context.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+channel.publish({
+  some: 'message'
+});
+```
+
+#### `channel.subscribe(onMessage)`
+
+* `onMessage` {Function} The handler to receive channel messages
+  * `message` {any} The message data
+  * `name` {string|symbol} The name of the channel
+
+Register a message handler to subscribe to this channel. This message handler
+will be run synchronously whenever a message is published to the channel. Any
+errors thrown in the message handler will trigger an [`'uncaughtException'`][].
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+channel.subscribe((message, name) => {
+  // Received data
+});
+```
+
+#### `channel.unsubscribe(onMessage)`
+
+* `onMessage` {Function} The previous subscribed handler to remove
+
+Remove a message handler previously registered to this channel with
+[`channel.subscribe(onMessage)`][].
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+function onMessage(message, name) {
+  // Received data
+}
+
+channel.subscribe(onMessage);
+
+channel.unsubscribe(onMessage);
+```
+
+#### `channel.span(data)`
+
+* `data` {any} The message to send as the start event of the span
+* Returns: {Span} The span object
+
+Create a span object to tie a sequence of events together. This wraps the input
+data in a new object which includes an action type of `start`, `annotation`, or
+`end` and a numeric id to link span messages together for later reconstitution.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+const channel = diagnostics_channel.channel('my-channel');
+
+const span = channel.span({
+  some: 'message'
+});
+```
+
+### Class: `Span`
```

Comment:
Couldn't a `Span` be a subclass of `Channel`? Doing so would allow nested spans, and would likely allow for a less complicated API surface

Reply:
Maybe, but then it'd need a separate API from `dc.channel(...)` to find or generate the differently-typed channel, and would probably need a separate `dc.hasSubscribers(name)` function, or at least one which is able to search both sets of typed channels. It would be very difficult to mix the types due to the data racing nature of `dc.channel(...)` between the publish and subscriber.

What's your thinking on how it could be nested? The design I have for spans currently is structured the way it is mainly to demonstrate the concept in a way that is easily removable if we decide it doesn't adequately match our needs. I made it a separate commit for exactly that reason. Conceptually what we need is just a shared identifier _of some sort_ between messages to correlate them. How we achieve that is not super important. I've also been considering a more basic API to just generate process-unique incrementing IDs and letting the user decide how to add the ID to their messages. This would make the generic `Span.aggregate` no longer possible, but would simplify things somewhat. Another possibility, anyway.

Reply:
Marking as resolved as I've removed the span API from this PR, feel free to bring this up when I open the separate PR for the span API.